README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

[![Travis-CI Build
Status](https://travis-ci.org/anre005/pinbasic.svg?branch=master)](https://travis-ci.org/anre005/pinbasic)
[![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/pinbasic)](https://cran.r-project.org/package=pinbasic)
[![Research software
impact](http://depsy.org/api/package/cran/pinbasic/badge.svg)](http://depsy.org/package/r/pinbasic)

# pinbasic: Fast and Stable Estimation of the Probability of Informed Trading (PIN)

The `pinbasic` package ships utilities for fast and stable estimation of
the probability of informed trading in the static PIN framework. The
function design is chosen to fit the extended EHO model setup but can
also be applied to the simpler EKOP model by equating the intensities of
uninformed buys and sells. State-of-the-art factorization of the model
likelihood function as well as most recent algorithms for generating
initial values for optimization routines are implemented. In total, two
likelihood factorizations and three methodologies for starting values
are included. Likelihood functions are evaluated with `pin_ll` and sets
of starting values are returned by `initial_vals`. The probability of
informed trading can be estimated for arbitrary length of daily buys and
sells data with `pin_est` which is a wrapper around the workhorse
function `pin_est_core`. No information about the time span of the
underlying data is required to perform optimizations with `pin_est`.
However, the recommendation given in the literature is using at least
data for 60 trading days to ensure convergence of the likelihood
maximization. Quarterly estimates are returned by `qpin` which can be
visualized with `ggplot`. Datasets of daily aggregated numbers of buys
and sells can be simulated with `simulateBS`. Calculation of confidence
intervals for the probability of informed trading can be enabled by
`confint` argument in optimization routines (`pin_est_core`, `pin_est`
and `qpin`) or by calling `pin_confint` directly. Additionally,
posterior probabilities for conditions of trading days can be computed
with `posterior` and plotted with `ggplot`.

## Examples

The dataset `BSfrequent` cover 60 trading days and represent a
frequently traded equity. Model parameters and the probability of
informed trading can be estimated with `pin_est`.

``` r
library(pinbasic)

# Loading data
data("BSfrequent")

# Estimation
pin_freq <- pin_est(numbuys = BSfrequent[,"Buys"], numsells = BSfrequent[,"Sells"])
pin_freq
#> $Results
#>            Estimate  Std. error    t value      Pr(> t)
#> alpha        0.2000  0.05163873   3.873062 0.0001074766
#> delta        0.5000  0.14433702   3.464115 0.0005319794
#> epsilon_b 1805.4354  5.67230494 318.289553 0.0000000000
#> epsilon_s 1700.6753  5.50984189 308.661358 0.0000000000
#> mu         597.6107 14.60600145  40.915421 0.0000000000
#> 
#> $ll
#> loglike 
#> 1415607 
#> 
#> $pin
#>        PIN 
#> 0.03296587 
#> 
#> $conv
#> Convergence 
#>           0 
#> 
#> $message
#> [1] "relative convergence (4)"
#> 
#> $iterations
#> Iterations 
#>          1 
#> 
#> $init_vals
#>     alpha     delta epsilon_b epsilon_s        mu 
#>    0.2000    0.5000 1805.4259 1700.6852  597.6111 
#> 
#> $posterior
#>                 no         good          bad
#>  [1,] 1.000000e+00 1.054677e-28 8.603585e-38
#>  [2,] 1.000000e+00 1.708301e-41 5.437236e-31
#>  [3,] 1.000000e+00 7.029207e-40 1.412484e-38
#>  [4,] 2.119110e-45 2.484190e-80 1.000000e+00
#>  [5,] 4.284377e-34 1.000000e+00 8.485090e-66
#>  [6,] 1.000000e+00 2.173042e-38 3.616880e-32
#>  [7,] 1.000000e+00 1.657474e-39 2.486257e-44
#>  [8,] 6.148468e-35 8.501740e-62 1.000000e+00
#>  [9,] 1.000000e+00 1.584055e-36 7.734248e-39
#> [10,] 1.000000e+00 6.617151e-36 9.224399e-43
#> [11,] 3.419344e-38 1.000000e+00 2.743870e-70
#> [12,] 1.000000e+00 1.264228e-40 9.395923e-40
#> [13,] 1.000000e+00 1.785679e-25 3.020421e-45
#> [14,] 1.000000e+00 6.820067e-38 3.359919e-44
#> [15,] 1.000000e+00 3.624041e-34 7.734248e-39
#> [16,] 6.036227e-38 1.015713e-77 1.000000e+00
#> [17,] 1.000000e+00 1.226614e-38 2.922883e-34
#> [18,] 1.000000e+00 5.443069e-42 2.276597e-42
#> [19,] 1.000000e+00 9.928515e-25 6.952753e-40
#> [20,] 3.917168e-47 1.000000e+00 1.103522e-86
#> [21,] 1.435856e-43 1.121835e-89 1.000000e+00
#> [22,] 1.000000e+00 1.632631e-38 1.908825e-38
#> [23,] 1.000000e+00 2.076785e-35 2.486257e-44
#> [24,] 1.540415e-42 1.000000e+00 1.148294e-86
#> [25,] 1.000000e+00 6.324039e-33 1.908825e-38
#> [26,] 1.000000e+00 2.892333e-38 3.737577e-43
#> [27,] 1.000000e+00 7.029207e-40 4.023421e-31
#> [28,] 1.000000e+00 9.642821e-42 2.627545e-35
#> [29,] 1.000000e+00 9.498287e-41 4.540581e-44
#> [30,] 1.000000e+00 1.560312e-35 2.235037e-45
#> [31,] 1.789734e-33 1.000000e+00 5.903585e-80
#> [32,] 1.472181e-39 1.000000e+00 7.064422e-74
#> [33,] 1.000000e+00 1.322823e-43 2.405970e-33
#> [34,] 1.000000e+00 2.764215e-35 3.192063e-36
#> [35,] 1.000000e+00 1.024329e-45 4.313742e-36
#> [36,] 1.000000e+00 2.045657e-34 9.395923e-40
#> [37,] 1.000000e+00 3.463511e-31 5.144867e-40
#> [38,] 1.000000e+00 1.632631e-38 1.412484e-38
#> [39,] 1.000000e+00 5.201964e-39 8.446526e-41
#> [40,] 1.000000e+00 4.472796e-29 1.520508e-26
#> [41,] 1.000000e+00 1.632631e-38 1.944319e-35
#> [42,] 1.000000e+00 3.072441e-42 3.550854e-35
#> [43,] 1.000000e+00 2.848980e-37 7.734248e-39
#> [44,] 1.000000e+00 2.108388e-36 6.825829e-43
#> [45,] 1.000000e+00 2.108388e-36 2.154180e-51
#> [46,] 1.000000e+00 1.513888e-33 8.024552e-33
#> [47,] 1.000000e+00 6.617151e-36 1.600465e-34
#> [48,] 1.000000e+00 2.108388e-36 3.020421e-45
#> [49,] 1.000000e+00 1.120353e-32 7.213723e-34
#> [50,] 2.401133e-41 4.101858e-82 1.000000e+00
#> [51,] 1.000000e+00 1.103560e-31 7.734248e-39
#> [52,] 1.000000e+00 7.355002e-43 1.684626e-42
#> [53,] 1.000000e+00 5.695348e-45 8.446526e-41
#> [54,] 1.000000e+00 5.201964e-39 1.514405e-43
#> [55,] 1.000000e+00 4.897022e-35 1.571248e-37
#> [56,] 1.000000e+00 6.923847e-39 1.293372e-36
#> [57,] 1.000000e+00 6.617151e-36 1.341919e-30
#> [58,] 1.000000e+00 1.172280e-35 3.877866e-37
#> [59,] 5.426305e-39 3.757091e-77 1.000000e+00
#> [60,] 1.000000e+00 2.486910e-28 2.084616e-40
#> attr(,"class")
#> [1] "matrix"    "posterior"
```

`BSfrequent2015` contains simulated daily buys and sells for a
frequently traded equity for business days in 2015. `qpin` returns
quarterly estimates which can be visualized with `ggplot`.

``` r
# Quarterly PIN estimates
# Confidence interval computation enabled:
#   * using only 1000 simulated datasets
#   * confidence level set to 0.95
#   * seed set to 123

data('BSfrequent2015')
qpin2015 <- qpin(numbuys = BSfrequent2015[,"Buys"], numsells = BSfrequent2015[,"Sells"],
                 dates = as.Date(rownames(BSfrequent2015), format = "%Y-%m-%d"),
                 confint = TRUE, ci_control = list(n = 1000, seed = 123))

# Print confidence intervals for all four quarters
ci_quarters <- lapply(qpin2015[["res"]], function(x) x$confint)
ci_quarters
#> $`2015.1`
#>       2.5%      97.5% 
#> 0.02752256 0.05943952 
#> 
#> $`2015.2`
#>        2.5%       97.5% 
#> 0.006405346 0.033145972 
#> 
#> $`2015.3`
#>       2.5%      97.5% 
#> 0.03658681 0.07322781 
#> 
#> $`2015.4`
#>       2.5%      97.5% 
#> 0.01014942 0.03794254

# Visualization of estimated parameters
library(ggplot2)
ggplot(qpin2015[["res"]])
```

![](tools/unnamed-chunk-3-1.png)<!-- -->

Posterior probabilities of trading days’ condition are returned by
`posterior` and can be displayed with `ggplot`. The following code chunk
shows how posterior probabilities for `BSfrequent2015` in the third
quarter can be calculated and visualized.

``` r
# Corresponding parameter estimates
freq_2015.3 <- qpin2015[["res"]]$'2015.3'$Results[,"Estimate"]

# Subsetting data
third_quarter <- subset(BSfrequent2015, subset = lubridate::quarter(rownames(BSfrequent2015)) == 3)

# Calculating posterior probabilities
post_third <- posterior(param = freq_2015.3, 
                        numbuys = third_quarter[,"Buys"], numsells = third_quarter[,"Sells"])

# Plotting
ggplot(post_third)
```

![](tools/postdates-1.png)<!-- -->

## Installation

You may install the stable version from **CRAN**, or the development
version from GitHub using **devtools**:

``` r
# install from CRAN
install.packages("pinbasic")

# install from github using devtools
devtools::install_github("anre005/pinbasic")
```


NEWS
pinbasic v1.2.3
----------------------------------------------------------------

Fixes:

  * Fix check in if statement (length > 1) which become an error in next R versions

pinbasic v1.2.1 (Release Date: 2018-11-18)
----------------------------------------------------------------

Fixes:

  * Removed detritus from vignette directory


pinbasic v1.2.1 (Release Date: 2018-08-19)
----------------------------------------------------------------

Changes:

  * added argument `nlminb_control` to function `qpin`, `pin_est` and `pin_est_core` (see control argument for `stats::nlminb`)
  * Dataset `BSfrequent` is now simulated with the parameter set (0.2, 0.5, 2200, 2000, 800) instead of (0.2, 0.5, 1800, 1700, 600)


pinbasic v1.2.0 (Release Date: 2017-12-21)
----------------------------------------------------------------

Changes:

  * `simulateBS` now completely in C++
  * Method for generating sets of initial values in `pin_confint` function
    is no longer tied to `HAC`. Other available methods can be chosen via
    the `method` argument of the function. (Defaults to `"HAC"`)
  * additional boolean flag `posterior` to enable computing posterior probabilities of
    conditions of trading days for `pin_est`, `pin_est_core` and `qpin` functions
  * `qpin` function returns now list of list by default with slots `res` and `posterior`:
    `res` is relevant for `ggplot.qpin`


pinbasic v1.1.0 (Release Date: 2017-03-01)
----------------------------------------------------------------

Fixes:

  * `pin_confint` for `ncores = 1`; `seed` is set once at the beginning, hence results differ slightly from them of earlier version

Changes:

  * Code clean-up for `pin_confint` and thus less dependencies; `ncores = 1` by default now
  * Confidence interval computations invoked by `pin_est_core`, `pin_est` or `qpin` use only 1 CPU core by default now
  * Likelihood factorizations are re-implemented with `Rcpp`
  * `simulateBS` function is re-implemented with `Rcpp`

pinbasic v1.0.1 (Release Date: 2017-01-11)
----------------------------------------------------------------

Fixes:

  * fixed NOTES reported by CRAN checks



pinbasic v1.0.0 (Release Date: 2017-01-09)
----------------------------------------------------------------

New Functions:

  * `posterior`: calculates posterior probabilities for trading days' conditions
  * `ggplot.posterior`: Method for `ggplot` function for handling objects with class 'posterior' (Results from `posterior` function)

Changes:

  * `pin_confint`: using `iter` function from `iterators` package to iterate over list of simulated datasets if executed in parallel
                   which gains little speed-up in execution time
  * `qpin_plot` changed to `ggplot.qpin`: Method for `ggplot` function for handling objects with class 'qpin' (Results from `qpin` function)

Fixes:

  * fixed missing convergence code in optimization routines if `num_best_res = 1`



pinbasic v0.2.0 (Release Date: 2016-12-02)
----------------------------------------------------------------

New Functions:

  * simulateBS() for simulating daily buys and sells data

  * pin_confint(): computes confidence intervals for the probability of informed trading
    Changes

Changes:

  * `pin_est_core`, `pin_est` and `qpin` gained two new arguments: `confint` and `ci_control`

  * updated plotting structure for qpin_plot, facets are now grouped by probability parameters,
    intensity parameters and the probability of informed trading

  * initial_vals together with method = "HAC_Ref" now returns a number of sets of initial values
    depending on num_clust argument, not only one set

  * Vignette was added



pinbasic v0.1.0 (Release Date: 2016-10-25)
----------------------------------------------------------------

* initial release

R/pin_max.R
#' Optimization in PIN models
#'
#' Core function for optimization routines
#'
#' Grants the most control over optimization procedure.
#' User-friendly wrappers for estimation with trading data of arbitrary length and quarterly data
#' are implemented with \code{\link{pin_est}} and \code{\link{qpin}}, respectively.
#' \code{\link{nlminb}} function in the \pkg{stats} package is used for maximization.
#' Vectors for \code{numbuys} and \code{numsells} need to have same length. \cr \cr
#' Confidence intervals for the probability of informed trading are calculated via Monte-Carlo-Simulation
#' if \code{confint = TRUE}. Settings of the confidence interval simulation can be specified via a named list for \code{ci_control}.
#' Valid list names are \code{n}, \code{seed}, \code{level} and \code{ncores} which pass
#' number of simulation runs (defaults to 10000), seed for RNG (defaults to \code{seed = NULL}),
#' confidence level (defaults to 0.95) and number of cpu cores utilized (defaults to 1).
#'
#' @inheritParams pin_ll
#' @param init_vals \emph{numeric}: matrix of initial values: either generated by \code{\link{initial_vals}} or
#'                   user-defined matrix with five columns for which colnames must consist of:
#'                   \code{alpha}, \code{delta}, \code{epsilon_b}, \code{epsilon_s}, \code{mu}
#' @param factorization \emph{character}: factorization of likelihood function: either 'EHO' or 'Lin_Ke', defaults to: 'Lin_Ke'
#' @param lower \emph{numeric}: lower bounds for optimization, must have length of 5
#' @param upper \emph{numeric}: upper bounds for optimization, must have length of 5
#' @param num_best_res Number of optimization runs for which results should be returned, either numeric or 'all',
#'                     only relevant if \code{init_vals} = 'Grid', defaults to 1
#' @param only_converged \emph{logical}: Return only results for which the likelihood converged?
#'                       Defaults to \code{TRUE}
#' @param nlminb_control \emph{list}: Control list for \code{\link[stats]{nlminb}}
#' @param confint \emph{logical}: Compute confidence intervals for PIN?
#'                 Defaults to \code{FALSE}
#' @param ci_control \emph{list}: see \strong{Details}
#' @param posterior \emph{logical}: Should posterior probabilities for conditions of trading days be computed?
#'
#' @seealso \code{\link{nlminb}},
#'          \code{\link{initial_vals}}
#'          \code{\link{pin_est}}
#'          \code{\link{qpin}}
#'          \code{\link{pin_confint}}
#'
#' @importFrom parallel detectCores
#' @importFrom stats nlminb
#'
#' @references
#' Easley, David et al. (2002) \cr
#' Is Information Risk a Determinant of Asset Returns? \cr
#' \emph{The Journal of Finance}, Volume 57, Number 5, pp. 2185 - 2221 \cr
#' \doi{10.1111/1540-6261.00493}
#'
#' Easley, David et al. (1996) \cr
#' Liquidity, Information, and Infrequently Traded Stocks\cr
#' \emph{The Journal of Finance}, Volume 51, Number 4, pp. 1405 - 1436 \cr
#' \doi{10.1111/j.1540-6261.1996.tb04074.x}
#'
#' Easley, David et al. (2010) \cr
#' Factoring Information into Returns \cr
#' \emph{Journal of Financial and Quantitative Analysis}, Volume 45, Issue 2, pp. 293 - 309 \cr
#' \doi{10.1017/S0022109010000074}
#'
#' Ersan, Oguz and Alici, Asli (2016) \cr
#' An unbiased computation methodology for estimating the probability of informed trading (PIN) \cr
#' \emph{Journal of International Financial Markets, Institutions and Money}, Volume 43, pp. 74 - 94 \cr
#' \doi{10.1016/j.intfin.2016.04.001}
#'
#' Gan, Quan et al. (2015) \cr
#' A faster estimation method for the probability of informed trading
#' using hierarchical agglomerative clustering \cr
#' \emph{Quantitative Finance}, Volume 15, Issue 11, pp. 1805 - 1821 \cr
#' \doi{10.1080/14697688.2015.1023336}
#'
#' Lin, Hsiou-Wei William and Ke, Wen-Chyan (2011) \cr
#' A computing bias in estimating the probability of informed trading \cr
#' \emph{Journal of Financial Markets}, Volume 14, Issue 4, pp. 625 - 640 \cr
#' \doi{10.1016/j.finmar.2011.03.001}
#'
#' Revolution Analytics and Steve Weston (2015) \cr
#' doParallel: Foreach Parallel Adaptor for the 'parallel' Package \cr
#' \emph{R package version 1.0.10}
#'
#' Revolution Analytics and Steve Weston (2015) \cr
#' foreach: Provides Foreach Looping Construct for R \cr
#' \emph{R package version 1.4.3}

#' Yan, Yuxing and Zhang, Shaojun (2012) \cr
#' An improved estimation method and empirical properties of the probability of informed trading \cr
#' \emph{Journal of Banking & Finance}, Volume 36, Issue 2, pp. 454 - 467 \cr
#' \doi{10.1016/j.jbankfin.2011.08.003}
#'
#' @return
#' If \code{num_best_res} = 1, a list with following elements is returned:
#' \describe{
#' \item{Results}{Matrix containing the parameter estimates as well as their estimated standard errors,
#'  t-values and p-values.}
#'  \item{ll}{Value of likelihood function returned by \code{nlminb}}
#'  \item{pin}{Estimated probability of informed trading}
#'  \item{conv}{Convergence code for nlminb optimization}
#'  \item{message}{Convergence message returned by the nlminb optimizer}
#'  \item{iterations}{Number of iterations until convergence of nlminb optimizer}
#'  \item{init_vals}{Vector of initial values}
#'  \item{confint}{If \code{confint = TRUE}; confidence interval for the probability of informed trading}
#'  }
#'
#'  If \code{num_best_res} > 1, a named list of lists is returned. Each component of the outer list
#'  is again a list structured as shown above.
#'  Naming scheme for the outer list is 'Best1',..., 'Best\code{num_best_res}'.
#'
#' @examples
#' # Loading simulated data for frequently traded stock
#'
#' data("BSfrequent")
#'
#' # Generate several matrices of initial values utilizing all methods implemented
#'
#' inits_grid <- initial_vals(numbuys = BSfrequent[,"Buys"],
#'                            numsells = BSfrequent[,"Sells"],
#'                            method = "Grid")
#'
#' inits_hac <- initial_vals(numbuys = BSfrequent[,"Buys"],
#'                           numsells = BSfrequent[,"Sells"],
#'                           method = "HAC")
#'
#' inits_hac_ref <- initial_vals(numbuys = BSfrequent[,"Buys"],
#'                               numsells = BSfrequent[,"Sells"],
#'                               method = "HAC_Ref")
#'
#' # Optimization with different matrices of initial values
#'
#' pin_core_grid <- pin_est_core(numbuys = BSfrequent[,"Buys"],
#'                               numsells = BSfrequent[,"Sells"],
#'                               factorization = "Lin_Ke", init_vals = inits_grid,
#'                               lower = rep(0,5), upper = c(1,1, rep(Inf,3)),
#'                               num_best_res = 5)
#'
#' pin_core_hac <- pin_est_core(numbuys = BSfrequent[,"Buys"],
#'                              numsells = BSfrequent[,"Sells"],
#'                              factorization = "Lin_Ke", init_vals = inits_hac,
#'                              lower = rep(0,5), upper = c(1,1, rep(Inf,3)))
#'
#' pin_core_hac_ref <- pin_est_core(numbuys = BSfrequent[,"Buys"],
#'                                  numsells = BSfrequent[,"Sells"],
#'                                  factorization = "Lin_Ke", init_vals = inits_hac_ref,
#'                                  lower = rep(0,5), upper = c(1,1, rep(Inf,3)))
#'
#' \dontrun{
#' pin_core_hac <- pin_est_core(numbuys = BSfrequent[,"Buys"],
#'                              numsells = BSfrequent[,"Sells"],
#'                              factorization = "Lin_Ke", init_vals = inits_hac,
#'                              lower = rep(0,5), upper = c(1,1, rep(Inf,3)),
#'                              confint = TRUE)
#' }
#' @export pin_est_core

pin_est_core <- function(numbuys = NULL, numsells = NULL,
                         factorization = "Lin_Ke",
                         init_vals = NULL, lower = rep(0,5), upper = c(1,1,rep(Inf,3)),
                         num_best_res = 1, only_converged = TRUE,
                         nlminb_control = list(),
                         confint = FALSE, ci_control = list(), posterior = TRUE) {
  if(is.null(init_vals)) stop("No initial values provided!")
  if(is.null(lower) || is.null(upper)) stop("Lower or upper bounds missing!")
  if(length(numbuys) != length(numsells)) stop("Unequal lengths for 'numbuys' and 'numsells'")

  factr <- match.arg(factorization, choices = c("Lin_Ke", "EHO"))

  mat <- matrix(data = NA, nrow = nrow(init_vals), ncol = ncol(init_vals) + 4)
  colnames(mat) <- c(colnames(init_vals), "loglike", "PIN", "Convergence", "Iterations")

  opt_message <- numeric(nrow(init_vals))

  fn <- function(x) pin_ll(param = x,
                           numbuys = numbuys, numsells = numsells,
                           factorization = factr)

  par_names <- c("alpha", "delta", "epsilon_b", "epsilon_s", "mu")

  if(nrow(mat) == 1) num_best_res <- 1

  ci_con <- list(n = 10000, seed = NULL, level = 0.95, ncores = 1)
  names_ci <- names(ci_con)

  ci_con[(nam_ci <- names(ci_control))] <- ci_control

  if (length(noNms <- nam_ci[!nam_ci %in% names_ci]))
    warning("unknown names in control: ", paste(noNms, collapse = ", "))
  if(ci_con$ncores < 1)
    stop("Set valid number of cpu cores for 'ncores'")

  nlminb_con <- list(eval.max = 1000,
                     iter.max = 500,
                     trace = 0,
                     abs.tol = 0,
                     rel.tol = 1e-10,
                     x.tol = 1.5e-8,
                     xf.tol = 2.2e-14,
                     step.min = 1, step.max = 1,
                     sing.tol = 1e-10)
  names_nlminb <- names(nlminb_con)
  nlminb_con[(nam_nlminb <- names(nlminb_control))] <- nlminb_control

  if (length(noNms_nlminb <- nam_nlminb[!nam_nlminb %in% names_nlminb]))
    warning("unknown names in control: ", paste(noNms_nlminb, collapse = ", "))

  for(i in 1:nrow(mat)) {
    tmp <- nlminb(start = init_vals[i,], objective = function(x) -fn(x),
                  lower = lower, upper = upper,
                  control = nlminb_con)

    mat[i, par_names] <- tmp$par
    mat[i,"loglike"] <- -tmp$objective
    mat[i,"PIN"] <- pin_calc(param = tmp$par)
    mat[i,"Convergence"] <- as.integer(tmp$convergence)
    mat[i,"Iterations"] <- as.integer(tmp$iterations)
    opt_message[i] <- tmp$message
  }
  if(nrow(mat) > 1) {
    mat <- mat[order(mat[,"loglike"], decreasing = TRUE),]
    start_vals <- init_vals[order(mat[,"loglike"], decreasing = TRUE),]
    if(only_converged) {
      # Convergence code 0 means succesful optimization
      converged <- as.logical(!mat[,"Convergence"])
      mat <- mat[converged,]
      start_vals <- start_vals[converged,]
    }
  }
  else start_vals <- matrix(init_vals[1,], nrow = 1)
  # num_best_res sets the number of returned parameter sets
  if(num_best_res == "all") num_best_res <- nrow(mat)

  if(num_best_res > 1) {
    mat_list <- vector("list", num_best_res)
    names(mat_list) <- paste0("Best", 1:num_best_res)
    for(i in 1:num_best_res) {
      mat_list[[paste0("Best",i)]] <- vector("list", 7)
      names(mat_list[[paste0("Best",i)]]) <- c("Results",
                                               "ll",
                                               "pin",
                                               "conv",
                                               "message",
                                               "iterations",
                                               "init_vals")
      if(confint) mat_list[[paste0("Best",i)]][["confint"]] <- numeric(2)

      tmp_res <- summary_car(param = mat[i,par_names],
                             numbuys = numbuys, numsells = numsells,
                             factorization = factorization,
                             lower = lower, upper = upper)
      if(!is.null(tmp_res)) {
        mat_list[[paste0("Best",i)]][["Results"]] <- tmp_res
      } else {
        mat_list[[paste0("Best",i)]][["Results"]] <- matrix(data = NA, ncol = 4, nrow = 5)
        colnames(mat_list[[paste0("Best",i)]][["Results"]]) <- c("Estimate", "Std. error", "t value", "Pr(> t)")
        rownames(mat_list[[paste0("Best",i)]][["Results"]]) <- c("alpha", "delta", "epsilon_b", "epsilon_s", "mu")
        mat_list[[paste0("Best",i)]][["Results"]][,"Estimate"] <- mat[i,par_names]
      }

      mat_list[[paste0("Best",i)]][["ll"]] <- mat[i,"loglike"]
      mat_list[[paste0("Best",i)]][["pin"]] <- mat[i,"PIN"]
      mat_list[[paste0("Best",i)]][["conv"]] <- mat[i,"Convergence"]
      mat_list[[paste0("Best",i)]][["message"]] <- opt_message[i]
      mat_list[[paste0("Best",i)]][["iterations"]] <- mat[i,"Iterations"]
      mat_list[[paste0("Best",i)]][["init_vals"]] <- start_vals[i,]
      if(confint) {
        mat_list[[paste0("Best",i)]][["confint"]] <- pin_confint(param = mat[i,par_names],
                                                                 numbuys = numbuys, numsells = numsells,
                                                                 lower = lower, upper = upper,
                                                                 n = ci_con$n, seed = ci_con$seed,
                                                                 level = ci_con$level, ncores = ci_con$ncores)
      }
      if(posterior) {
        mat_list[[paste0("Best",i)]][["posterior"]] <- posterior(param = mat[i,par_names],
                                                                 numbuys = numbuys, numsells = numsells)
      }
    }
  } else {
    mat_list <- vector("list", 7)
    names(mat_list) <- c("Results",
                         "ll",
                         "pin",
                         "conv",
                         "message",
                         "iterations",
                         "init_vals")
    if(confint) mat_list[["confint"]] <- numeric(2)

    tmp_res <- summary_car(param = mat[1,par_names],
                           numbuys = numbuys, numsells = numsells,
                           factorization = factorization,
                           lower = lower, upper = upper)
    if(!is.null(tmp_res)) {
      mat_list[["Results"]] <- tmp_res
    } else {
      mat_list[["Results"]] <- matrix(data = NA, ncol = 4, nrow = 5)
      colnames(mat_list[["Results"]]) <- c("Estimate", "Std. error", "t value", "Pr(> t)")
      rownames(mat_list[["Results"]]) <- c("alpha", "delta", "epsilon_b", "epsilon_s", "mu")
      mat_list[["Results"]][,"Estimate"] <- mat[i,par_names]
    }

    mat_list[["ll"]] <- mat[1,"loglike"]
    mat_list[["pin"]] <- mat[1,"PIN"]
    mat_list[["conv"]] <- mat[1,"Convergence"]
    mat_list[["message"]] <- opt_message[1]
    mat_list[["iterations"]] <- mat[1,"Iterations"]
    mat_list[["init_vals"]] <- start_vals[1,]
    names(mat_list[["init_vals"]]) <- c("alpha", "delta", "epsilon_b", "epsilon_s", "mu")
    if(confint) {
      mat_list[["confint"]] <- pin_confint(param = mat[1,par_names],
                                           numbuys = numbuys, numsells = numsells,
                                           lower = lower, upper = upper,
                                           n = ci_con$n, seed = ci_con$seed,
                                           level = ci_con$level, ncores = ci_con$ncores)
    }
    if(posterior) {
      mat_list[["posterior"]] <- posterior(param = mat[1,par_names],
                                           numbuys = numbuys, numsells = numsells)
    }
  }
  mat_list
}
